internal/factsengine/gatherers/sudoers.go

```diff
@@ -0,0 +1,298 @@
+
+		if factReq.Argument == "" {
+			result, err := g.gatherAll(ctx)
+			if err != nil {
+				return nil, err
+			}
+			data = result
+		} else {
+			result, err := g.gatherSingle(ctx, factReq.Argument)
+			if err != nil {
+				return nil, err
+			}
+			data = []parsedSudoers{result}
+		}
+
+		value, err := toFactValue(data)
+		if err != nil {
+			return nil, err
+		}
+		fact = entities.NewFactGatheredWithRequest(factReq, value)
+		facts = append(facts, fact)
+	}
+
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	log.Infof("Requested %s facts gathered", OSReleaseGathererName)
+	return facts, nil
+}
+
+func (g *SudoersGatherer) gatherAll(ctx context.Context) ([]parsedSudoers, *entities.FactGatheringError) {
+	usernames, err := findUsernames(g.fs)
+	if err != nil {
+		return nil, SudoersUserError.Wrap(err.Error())
+	}
+	allUsers := []parsedSudoers{}
+	for _, username := range usernames {
+		single, err := g.gatherSingle(ctx, username)
+		if err != nil {
+			return nil, err
+		}
+		allUsers = append(allUsers, single)
+	}
+	return allUsers, nil
+}
+
+func (g *SudoersGatherer) gatherSingle(
+	ctx context.Context,
+	username string,
+) (parsedSudoers, *entities.FactGatheringError) {
+	output, err := g.readUserPrivileges(ctx, username)
+	if err != nil {
+		return parsedSudoers{}, SudoersReadError.Wrap(err.Error())
+	}
+
+	privileges, err := g.parseUserPrivileges(string(output))
+	if err != nil {
+		return parsedSudoers{}, SudoersParseError.Wrap(err.Error())
+	}
+
+	return parsedSudoers{
+		CommandsAsRoot: privileges,
+		User:           username,
+	}, nil
+
+}
+
+func (g *SudoersGatherer) readUserPrivileges(ctx context.Context, username string) ([]byte, error) {
+	err := validateUsername(username)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid username "+username)
+	}
+	output, err := g.executor.ExecContext(ctx, "/usr/bin/sudo", "-l", "-U", username)
+	if err != nil {
+		return nil, errors.Wrap(err, "error running sudo command")
+	}
+	return output, nil
+}
+
+func (g *SudoersGatherer) parseUserPrivileges(output string) ([]privilegeEntry, error) {
+	var privileges []privilegeEntry
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	var inPrivilegesSection bool
+
+	// This regex detects the header that indicates the start of the command privileges.
+	privilegesStartRegex := regexp.MustCompile(`^User\s+\S+\s+may run the following commands on`)
+	// This regex parses each privilege entry line.
+	// Group 1 captures the run-as specifier inside parentheses.
+	// Group 2 optionally captures "NOPASSWD:" or "PASSWD:".
+	// Group 3 captures the commands and any arguments.
+	entryRegex := regexp.MustCompile(`^\s*\(([^)]+)\)\s*(?:(NOPASSWD:|PASSWD:)\s*)?(.*)$`)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// Check for the header line to start processing privilege entries.
+		if privilegesStartRegex.MatchString(line) {
+			inPrivilegesSection = true
+			continue // Skip the header line itself.
+		}
+
+		if inPrivilegesSection {
```

Comment:
suggestion: I think we can use the early return technique to remove some indentation if we use it this way:
```
if !inPrivilegesSection {
    continue
}
```

Reply:
Done in https://github.com/trento-project/agent/pull/404/commits/57164b0bd6af97fc2a76db1eba42a3721d246c7b